File: Sources/CodexRadarCore/NotificationPolicy.swift
import Foundation

public struct NotificationEvent: Equatable {
    public enum Severity: String, Equatable {
        case passive
        case active
        case urgent
    }

    public let identifier: String
    public let title: String
    public let body: String
    public let severity: Severity

    public init(
        identifier: String,
        title: String,
        body: String,
        severity: Severity
    ) {
        self.identifier = identifier
        self.title = title
        self.body = body
        self.severity = severity
    }
}

public struct NotificationMemory: Equatable {
    public var initialized: Bool
    public var lastSpeedOpenKey: String?
    public var lastResetCloseKey: String?
    public var lastPredictionKey: String?
    public var lastIQKey: String?
    public var lastWeeklyWarningKey: String?
    public var lastWeeklyCriticalKey: String?
    public var lastWeeklyRestoreKey: String?
    public var lastWeeklyWarningAt: Date?
    public var lastWeeklyCriticalAt: Date?

    public init(
        initialized: Bool = false,
        lastSpeedOpenKey: String? = nil,
        lastResetCloseKey: String? = nil,
        lastPredictionKey: String? = nil,
        lastIQKey: String? = nil,
        lastWeeklyWarningKey: String? = nil,
        lastWeeklyCriticalKey: String? = nil,
        lastWeeklyRestoreKey: String? = nil,
        lastWeeklyWarningAt: Date? = nil,
        lastWeeklyCriticalAt: Date? = nil
    ) {
        self.initialized = initialized
        self.lastSpeedOpenKey = lastSpeedOpenKey
        self.lastResetCloseKey = lastResetCloseKey
        self.lastPredictionKey = lastPredictionKey
        self.lastIQKey = lastIQKey
        self.lastWeeklyWarningKey = lastWeeklyWarningKey
        self.lastWeeklyCriticalKey = lastWeeklyCriticalKey
        self.lastWeeklyRestoreKey = lastWeeklyRestoreKey
        self.lastWeeklyWarningAt = lastWeeklyWarningAt
        self.lastWeeklyCriticalAt = lastWeeklyCriticalAt
    }
}

public struct NotificationPolicy {
    public init() {}

    public func evaluate(
        previous: DashboardState?,
        current: DashboardState,
        memory: inout NotificationMemory,
        now: Date = Date()
    ) -> [NotificationEvent] {
        let wasInitialized = memory.initialized
        if !memory.initialized {
            seedInitialMemory(current: current, memory: &memory)
        }

        var events: [NotificationEvent] = []
        appendSpeedWindowEvents(
            current: current,
            wasInitialized: wasInitialized,
            memory: &memory,
            events: &events
        )
        guard wasInitialized else {
            return events
        }
        appendResetEvents(current: current, memory: &memory, events: &events)
        appendWeeklyLimitEvents(
            previous: previous,
            current: current,
            memory: &memory,
            events: &events,
            now: now
        )
        appendPredictionEvents(
            previous: previous,
            current: current,
            memory: &memory,
            events: &events
        )
        appendIQEvents(current: current, memory: &memory, events: &events)
        return events
    }

    private func seedInitialMemory(current: DashboardState, memory: inout NotificationMemory) {
        memory.initialized = true
        if let key = speedOpenKey(current: current), current.current?.windowOpen != true {
            memory.lastSpeedOpenKey = key
        }
        if let key = resetCloseKey(current: current) {
            memory.lastResetCloseKey = key
        }
        if let key = iqKey(current: current) {
            memory.lastIQKey = key
        }
        if let key = predictionKey(current: current) {
            memory.lastPredictionKey = key
        }
    }

    private func appendSpeedWindowEvents(
        current: DashboardState,
        wasInitialized: Bool,
        memory: inout NotificationMemory,
        events: inout [NotificationEvent]
    ) {
        guard current.activeSpeedWindow, let key = speedOpenKey(current: current) else {
            return
        }
        guard memory.lastSpeedOpenKey != key else {
            return
        }
        memory.lastSpeedOpenKey = key
        let weekly = DisplayFormatters.percent(current.rateLimits?.weeklyRemainingPercent)
        events.append(NotificationEvent(
            identifier: "speed-window-open-\(key)",
            title: "速蹬窗口开启",
            body: wasInitialized ? "当前周额度剩余 \(weekly)，建议尽快使用。" : "App 启动时检测到窗口已开启，当前周额度剩余 \(weekly)。",
            severity: .urgent
        ))
    }

    private func appendResetEvents(
        current: DashboardState,
        memory: inout NotificationMemory,
        events: inout [NotificationEvent]
    ) {
        guard let key = resetCloseKey(current: current) else {
            return
        }
        guard memory.lastResetCloseKey != key else {
            return
        }
        memory.lastResetCloseKey = key
        let title = current.current?.lastWindow?.title ?? "Codex limit reset"
        events.append(NotificationEvent(
            identifier: "reset-close-\(key)",
            title: "CodexRadar 记录到 reset",
            body: title,
            severity: .urgent
        ))
    }

    private func appendWeeklyLimitEvents(
        previous: DashboardState?,
        current: DashboardState,
        memory: inout NotificationMemory,
        events: inout [NotificationEvent],
        now: Date
    ) {
        guard let remaining = current.rateLimits?.weeklyRemainingPercent else {
            return
        }
        let resetKey = current.rateLimits?.weeklyBucket?.resetsAt.map(String.init) ?? "unknown"

        if remaining <= AppConstants.criticalRemainingPercent {
            let key = "\(resetKey):critical"
            if shouldSendQuotaNotification(
                key: key,
                lastKey: memory.lastWeeklyCriticalKey,
                lastSentAt: memory.lastWeeklyCriticalAt,
                cooldown: AppConstants.weeklyCriticalNotificationCooldownSeconds,
                now: now
            ) {
                memory.lastWeeklyCriticalKey = key
                memory.lastWeeklyCriticalAt = now
                events.append(NotificationEvent(
                    identifier: "weekly-critical-\(key)",
                    title: "Codex 周额度很低",
                    body: "当前周额度剩余 \(remaining)%。",
                    severity: .urgent
                ))
            }
        } else if remaining <= AppConstants.warningRemainingPercent {
            let key = "\(resetKey):warning"
            if shouldSendQuotaNotification(
                key: key,
                lastKey: memory.lastWeeklyWarningKey,
                lastSentAt: memory.lastWeeklyWarningAt,
                cooldown: AppConstants.weeklyWarningNotificationCooldownSeconds,
                now: now
            ) {
                memory.lastWeeklyWarningKey = key
                memory.lastWeeklyWarningAt = now
                events.append(NotificationEvent(
                    identifier: "weekly-warning-\(key)",
                    title: "Codex 周额度偏低",
                    body: "当前周额度剩余 \(remaining)%。",
                    severity: .active
                ))
            }
        }

        guard let previousRemaining = previous?.rateLimits?.weeklyRemainingPercent else {
            return
        }
        let previousReset = previous?.rateLimits?.weeklyBucket?.resetsAt
        let currentReset = current.rateLimits?.weeklyBucket?.resetsAt
        guard previousRemaining <= AppConstants.warningRemainingPercent,
              remaining >= AppConstants.restoredRemainingPercent,
              previousReset != currentReset else {
            return
        }
        let key = "\(resetKey):restored"
        guard memory.lastWeeklyRestoreKey != key else {
            return
        }
        memory.lastWeeklyRestoreKey = key
        events.append(NotificationEvent(
            identifier: "weekly-restored-\(key)",
            title: "Codex 周额度已恢复",
            body: "当前周额度剩余 \(remaining)%。",
            severity: .active
        ))
    }

    private func shouldSendQuotaNotification(
        key: String,
        lastKey: String?,
        lastSentAt: Date?,
        cooldown: TimeInterval,
        now: Date
    ) -> Bool {
        guard lastKey != key else {
            return false
        }
        guard let lastSentAt else {
            return true
        }
        return now.timeIntervalSince(lastSentAt) >= cooldown
    }

    private func appendPredictionEvents(
        previous: DashboardState?,
        current: DashboardState,
        memory: inout NotificationMemory,
        events: inout [NotificationEvent]
    ) {
        guard current.current?.status?.lowercased() != "retired" else {
            return
        }
        guard let prediction = current.prediction else {
            return
        }
        let level = prediction.level?.lowercased()
        let previousLevel = previous?.prediction?.level?.lowercased()
        let shouldNotify = prediction.shouldNotify == true || (level == "high" && previousLevel != "high")
        guard shouldNotify, let key = predictionKey(current: current) else {
            return
        }
        guard memory.lastPredictionKey != key else {
            return
        }
        memory.lastPredictionKey = key
        let probability = prediction.probability24h.map { "\(Int(round($0 * 100)))%" } ?? "unknown"
        events.append(NotificationEvent(
            identifier: "prediction-\(key)",
            title: "Codex reset 预测升高",
            body: "未来 24h 概率 \(probability)，等级 \(prediction.level ?? "unknown")。",
            severity: .active
        ))
    }

    private func appendIQEvents(
        current: DashboardState,
        memory: inout NotificationMemory,
        events: inout [NotificationEvent]
    ) {
        guard let latest = current.modelIQ?.latest,
              let score = latest.iqScore,
              score < 80 || latest.status?.lowercased() == "red",
              let key = iqKey(current: current),
              memory.lastIQKey != key else {
            return
        }
        memory.lastIQKey = key
        let passed = latest.passed.map(String.init) ?? "?"
        let tasks = latest.tasks.map(String.init) ?? "?"
        let scoreText = DisplayFormatters.iqScore(score)
        events.append(NotificationEvent(
            identifier: "model-iq-\(key)",
            title: "Codex IQ 偏低",
            body: "IQ \(scoreText)，\(passed)/\(tasks) tasks 通过。",
            severity: .passive
        ))
    }

    private func speedOpenKey(current: DashboardState) -> String? {
        guard let window = current.current?.lastWindow else {
            return current.activeSpeedWindow ? "unknown-open" : nil
        }
        return "\(window.id ?? "unknown"):\(window.openedAt ?? "unknown")"
    }

    private func resetCloseKey(current: DashboardState) -> String? {
        guard let window = current.current?.lastWindow,
              window.status?.lowercased() == "closed",
              let closedAt = window.closedAt else {
            return nil
        }
        return "\(window.id ?? "unknown"):\(closedAt)"
    }

    private func predictionKey(current: DashboardState) -> String? {
        guard let prediction = current.prediction else {
            return nil
        }
        return "\(prediction.updatedAt ?? "unknown"):\(prediction.level ?? "unknown"):\(prediction.probability24h ?? -1)"
    }

    private func iqKey(current: DashboardState) -> String? {
        guard let latest = current.modelIQ?.latest else {
            return nil
        }
        return "\(latest.date ?? "unknown"):\(DisplayFormatters.iqScore(latest.iqScore)):\(latest.status ?? "unknown")"
    }
}

File: windows/CodexRadar.Windows/NotificationPolicy.cs
namespace CodexRadar.Windows;

internal enum NotificationSeverity { Passive, Active, Urgent }
internal sealed record NotificationEvent(string Identifier, string Title, string Body, NotificationSeverity Severity);

internal static class NotificationPolicy
{
    private const int WarningRemaining = 30;
    private const int CriticalRemaining = 15;
    private const int RestoredRemaining = 80;

    public static IReadOnlyList<NotificationEvent> Evaluate(
        DashboardSnapshot? previous, DashboardSnapshot current, AppSettings settings, DateTimeOffset? now = null)
    {
        var memory = settings.NotificationMemory;
        var wasInitialized = memory.Initialized;
        if (!memory.Initialized) Seed(current, memory);
        var events = new List<NotificationEvent>();
        AppendSpeed(current, wasInitialized, memory, settings.Chinese, events);
        if (!wasInitialized) return events;
        AppendReset(current, memory, settings.Chinese, events);
        AppendQuota(previous, current, memory, settings.Chinese, now ?? DateTimeOffset.Now, events);
        // Always advance the seen-event memory. Delivery preferences are applied by the tray
        // so re-enabling a category never replays an old prediction or IQ event.
        AppendPrediction(previous, current, memory, settings.Chinese, events);
        AppendIq(current, memory, settings.Chinese, events);
        return events;
    }

    private static void Seed(DashboardSnapshot current, NotificationMemory memory)
    {
        memory.Initialized = true;
        if (current.ResetCloseKey is { } reset) memory.LastResetCloseKey = reset;
        if (IqKey(current) is { } iq) memory.LastIqKey = iq;
        if (PredictionKey(current) is { } prediction) memory.LastPredictionKey = prediction;
    }

    private static void AppendSpeed(DashboardSnapshot state, bool initialized, NotificationMemory memory,
        bool zh, List<NotificationEvent> events)
    {
        if (state.SpeedAlertKey is not { } key || memory.LastSpeedOpenKey == key) return;
        memory.LastSpeedOpenKey = key;
        var weekly = state.WeeklyRemaining is int quota ? $"{quota}%" : "--";
        events.Add(new NotificationEvent($"speed-window-open-{key}", Zh(zh, "速蹬窗口开启", "Speed window open"),
            initialized ? Zh(zh, $"当前周额度剩余 {weekly}，建议尽快使用。", $"Weekly quota left: {weekly}. Use quota now.")
                : Zh(zh, $"启动时检测到窗口已开启，当前周额度剩余 {weekly}。", $"Window already open at launch; weekly quota left: {weekly}."),
            NotificationSeverity.Urgent));
    }

    private static void AppendReset(DashboardSnapshot state, NotificationMemory memory, bool zh, List<NotificationEvent> events)
    {
        if (state.ResetCloseKey is not { } key || memory.LastResetCloseKey == key) return;
        memory.LastResetCloseKey = key;
        events.Add(new NotificationEvent($"reset-close-{key}", Zh(zh, "CodexRadar 记录到 reset", "CodexRadar recorded a reset"),
            state.WindowTitle ?? "Codex limit reset", NotificationSeverity.Urgent));
    }

    private static void AppendQuota(DashboardSnapshot? previous, DashboardSnapshot state, NotificationMemory memory,
        bool zh, DateTimeOffset now, List<NotificationEvent> events)
    {
        if (state.WeeklyRemaining is not int remaining) return;
        var resetKey = state.WeeklyResetsAt?.ToUnixTimeSeconds().ToString() ?? "unknown";
        if (remaining <= CriticalRemaining)
        {
            var key = $"{resetKey}:critical";
            if (ShouldSend(key, memory.LastWeeklyCriticalKey, memory.LastWeeklyCriticalAt, TimeSpan.FromHours(4), now))
            {
                memory.LastWeeklyCriticalKey = key;
                memory.LastWeeklyCriticalAt = now;
                events.Add(new NotificationEvent($"weekly-critical-{key}", Zh(zh, "Codex 周额度很低", "Codex weekly quota is very low"),
                    Zh(zh, $"当前周额度剩余 {remaining}%。", $"Weekly quota left: {remaining}%."), NotificationSeverity.Urgent));
            }
        }
        else if (remaining <= WarningRemaining)
        {
            var key = $"{resetKey}:warning";
            if (ShouldSend(key, memory.LastWeeklyWarningKey, memory.LastWeeklyWarningAt, TimeSpan.FromHours(12), now))
            {
                memory.LastWeeklyWarningKey = key;
                memory.LastWeeklyWarningAt = now;
                events.Add(new NotificationEvent($"weekly-warning-{key}", Zh(zh, "Codex 周额度偏低", "Codex weekly quota is low"),
                    Zh(zh, $"当前周额度剩余 {remaining}%。", $"Weekly quota left: {remaining}%."), NotificationSeverity.Active));
            }
        }

        var restoreKey = state.WeeklyResetsAt is null ? null : $"{resetKey}:restored";
        if (memory.PendingWeeklyRestoreKey is { } pending)
        {
            if (pending == restoreKey && remaining >= RestoredRemaining)
            {
                memory.PendingWeeklyRestoreKey = null;
                if (memory.LastWeeklyRestoreKey != pending)
                {
                    memory.LastWeeklyRestoreKey = pending;
                    events.Add(new NotificationEvent($"weekly-restored-{pending}", Zh(zh, "Codex 周额度已恢复", "Codex weekly quota restored"),
                        Zh(zh, $"当前周额度剩余 {remaining}%。", $"Weekly quota left: {remaining}%."), NotificationSeverity.Active));
                }
                return;
            }
            if (pending != restoreKey || remaining < RestoredRemaining) memory.PendingWeeklyRestoreKey = null;
        }
        if (previous?.WeeklyRemaining <= WarningRemaining && remaining >= RestoredRemaining
            && previous.WeeklyResetsAt != state.WeeklyResetsAt && restoreKey is not null)
            memory.PendingWeeklyRestoreKey = restoreKey;
    }

    private static void AppendPrediction(DashboardSnapshot? previous, DashboardSnapshot state, NotificationMemory memory,
        bool zh, List<NotificationEvent> events)
    {
        if (state.RadarStatus?.Equals("retired", StringComparison.OrdinalIgnoreCase) == true || state.Prediction is not { } prediction) return;
        var level = prediction.Level?.ToLowerInvariant();
        var previousLevel = previous?.Prediction?.Level?.ToLowerInvariant();
        var shouldNotify = prediction.ShouldNotify == true || (level == "high" && previousLevel != "high");
        var key = PredictionKey(state);
        if (!shouldNotify || key is null || memory.LastPredictionKey == key) return;
        memory.LastPredictionKey = key;
        var probability = prediction.Probability24Hours is double value
            ? $"{Math.Round(NormalizeProbability(value), MidpointRounding.AwayFromZero):0}%"
            : "unknown";
        events.Add(new NotificationEvent($"prediction-{key}", Zh(zh, "Codex reset 预测升高", "Codex reset prediction increased"),
            Zh(zh, $"未来 24h 概率 {probability}，等级 {prediction.Level ?? "unknown"}。",
                $"24h probability {probability}; level {prediction.Level ?? "unknown"}."), NotificationSeverity.Active));
    }

    private static void AppendIq(DashboardSnapshot state, NotificationMemory memory, bool zh, List<NotificationEvent> events)
    {
        if (state.IqScore is not double score || (score >= 80 && !string.Equals(state.IqStatus, "red", StringComparison.OrdinalIgnoreCase))) return;
        var key = IqKey(state);
        if (key is null || memory.LastIqKey == key) return;
        memory.LastIqKey = key;
        events.Add(new NotificationEvent($"model-iq-{key}", Zh(zh, "Codex IQ 偏低", "Codex IQ is low"),
            $"IQ {score:0.0}, {state.Passed?.ToString() ?? "?"}/{state.ValidTasks?.ToString() ?? "?"} tasks.", NotificationSeverity.Passive));
    }

    private static bool ShouldSend(string key, string? previousKey, DateTimeOffset? sentAt, TimeSpan cooldown, DateTimeOffset now) =>
        previousKey != key && (sentAt is null || now - sentAt >= cooldown);
    private static string? PredictionKey(DashboardSnapshot state) => state.Prediction is not { } prediction ? null
        : $"{prediction.UpdatedAt?.ToUniversalTime().ToString("O") ?? "unknown"}:{prediction.Level ?? "unknown"}:{prediction.Probability24Hours ?? -1}";
    private static string? IqKey(DashboardSnapshot state) => state.IqScore is not double score ? null
        : $"{state.IqDate ?? "unknown"}:{score:0.0}:{state.IqStatus ?? "unknown"}";
    private static double NormalizeProbability(double value) => value <= 1 ? value * 100 : value;
    private static string Zh(bool chinese, string zh, string en) => chinese ? zh : en;
}
